python/cuml/benchmark/bench_all.py

```diff
@@ -0,0 +1,113 @@
+from cuml.benchmark import utils, bench_data
+from importlib import reload
+
+import sklearn.cluster, sklearn.neighbors
+import cuml.cluster, cuml.neighbors
+import umap
+import sys
+
+def get_all_algos(load_data):
+    return [
+        utils.AlgoComparisonWrapper(
```

Comment:
Mostly thinking out loud here. 

I like the simplicity behind this. I'm wondering if we should define these wrapper instances as globals somewhere, and whether we should specify different values/ranges of the `shared_args` to try, similar to `pytest.mark.parameter`

For example, it would be really neat if we had a global kmeans definition that we could just reuse everywhere else, like this:

```
kmeans_wrapper = utils.AlgoComparisonWrapper(
    sklearn.cluster.KMeans,
    cuml.cluster.KMeans,
    shared_args=dict(init = ["random", "kmeans||"], 
                                 n_clusters=np.logspace(2.0, 3.0, num=5), 
                                 max_iter=300]),
    name='kmeans')
) 
...
```

While this is being redesigned, I like the idea of abstracting the dataset away from the algorithm being benchmarked so that the benchmark being run can take care of loading the data, converting to the proper format, and running it through an `AlgoComparisonWrapper`.

Perhaps there could be a `Dataset` class that can be constructed w/ an array of data, which contains the relevant conversion functions. 

```
class Dataset(object):
    def load_data(n_rows, n_dims, out_type = 'cudf'):
        pass

class HiggsDataset(Dataset):
    def __init__(fname = "..."):
        # read higgs dataset into some internal structure

    def load_data(n_rows, n_dims, out_type = 'cudf'):
        return X or X,y depending on the type of dataset. 
```        
    

It would be neat if the interactions between objects could be something like below:

```
datasets = [HiggsDataset(fname=".."), BlobsDataset(n_clusters=5)...]

br = BenchmarkRunner(
    benchmarks = [utils.SpeedupBenchmark],
    data_types = ['cudf', 'numpy'],
    verbose = true,
    bench_rows = ...,
    bench_dims = ...
    n_trials = 3
)

results = BenchmarkResults()
for d in datasets:
    for a in algowrappers:
        results.combine(br.run(a, d,))

print(results.to_dataframe())
results.to_csv("results.csv")

BenchmarkChart(results).chart_sizes_by_dims().show()
```

This way, the benchmark runner is responsible for orchestrating the interactions between instances of `Benchmark`, `AlgoComparisonWrapper`, and `Dataset`

I also think the chart drawing should get separated from the `BenchmarkRunner` class. Perhaps the benchmark runner could produce an instance of  a `BenchmarkResults` class, which would know how store/load itself to/from disk (as csv, pickle, whatever). Default behavior would be for it to just to store a dict/dataframe in memory w/ the results of the benchmark:

```
class DataPoint(object):
    def __init__(self,n_rows, n_dims, algo_name, shared_args, dict(time_sk=..., time_cuml=..., speedup=...)

class BenchmarkResults(object):
    def combine(other_benchmark_results)
    def add_result(data_point):
        # Adds this entry to whatever data struct it is using internally to store the result of a single benchmark run. 

    def to_csv(self)
    def to_dataframe(self)
    def __setstate__(self)
    def __getstate__(self)
```

Charting could be its own beast, which could take a `BenchmarkResults` instance and draw different types of charts:

```
class BenchmarkChart(object):
    def __init__(self, benchmark_results)

    # The chart currently in the code
    def chart_sizes_by_dims(algo_name, shared_args, min_n_rows=..., max_n_rows=..., min_n_dims=..., max_n_dims=...)

    # Compares the shared args for a single algorithm
    def chart_compare_shared_args(algo_name, min_n_rows=..., max_n_rows=..., min_n_dims=...)

    # Compares algo times across several algorithms (algo_names=None compares all algos in result set), maybe aggregating across different shared args results. 
    def chart_compare_algo_times(algo_names = None, min_n_rows=..., max_n_rows=..., min_n_dims=...)
```

What do you think?

Reply:
> are you thinking that a dataset should include the nrows and nfeatures? 

Sure. Even with fixed-sized datasets we could sample uniformly to get any number of rows/features we need for a particular volume. Or perhaps set reasonable defaults that automatically use the default size of the dataset. 

I think the pytest parametrization is a great idea!